### PR TITLE
#2820034 Add item to a fresh cart fails if Editor module is enabled

### PR DESCRIPTION
**Pull request by agoradesign**
let the test fail...

### CODE REVIEW

**Comment by mglaman** [comment]
@agoradesign know what's weird? The tests pass for PHP7. So something about the "check if this should be refreshed" is wrong in different ways than we may think.

I really wish core had a `::isSaving()` method.

**Comment by agoradesign** [comment]
That's really weird! Yes, isSaving() would be great.. but as mentioned in the original issue on d.o, I'm suspecting that the OrderRefresh is working on the loaded $entity->original too and trying to save that. This would be the bigger problem

File: modules/cart/tests/modules/commerce_cart_test/commerce_cart_test.module [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+use \Drupal\Core\Entity\EntityInterface;
+
+/**
+ * Implements hook_entity_update().
+ *
+ * To replicate editor.module and catch save inceptions.
+ */
+function commerce_cart_test_entity_update(EntityInterface $entity) {
+  // Only act on order entities.
+  if (!($entity instanceof OrderInterface)) {
```
Comment (bojanz) [inline modules/cart/tests/modules/commerce_cart_test/commerce_cart_test.module]: Why not implement hook_commerce_order_update() then?

File: modules/cart/tests/modules/commerce_cart_test/commerce_cart_test.module [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    return;
+  }
+  if (!$entity->isNew() && empty($entity->original)) {
+    throw new \Exception('Found stale data! $entity->original must be present on updated entity!');
```
Comment (bojanz) [inline modules/cart/tests/modules/commerce_cart_test/commerce_cart_test.module]: This doesn't have to mean that the data is stale. Better to say '$entity->original not found', and perhaps make that exception EntityMalformedException?

File: modules/cart/tests/modules/commerce_cart_test/commerce_cart_test.module (hunk 1 above)
Comment (agoradesign) [reply modules/cart/tests/modules/commerce_cart_test/commerce_cart_test.module]: right... ok I'll update again...

File: modules/cart/tests/modules/commerce_cart_test/commerce_cart_test.module (hunk 2 above)
Comment (agoradesign) [reply modules/cart/tests/modules/commerce_cart_test/commerce_cart_test.module]: PR is updated with new solution

File: modules/order/src/OrderStorage.php [hunk 3]
```diff
@@ -44,6 +51,7 @@ class OrderStorage extends CommerceContentEntityStorage {
   public function __construct(EntityTypeInterface $entity_type, Connection $database, EntityManagerInterface $entity_manager, CacheBackendInterface $cache, LanguageManagerInterface $language_manager, EventDispatcherInterface $event_dispatcher, OrderRefreshInterface $order_refresh) {
     parent::__construct($entity_type, $database, $entity_manager, $cache, $language_manager, $event_dispatcher);
     $this->orderRefresh = $order_refresh;
+    $this->unchangedLoadIds = [];
```
Comment (bojanz) [inline modules/order/src/OrderStorage.php:54]: I just realized we can use a single flag for this, we don't need to track individual ids, they're processed one by one anyway. skipRefresh perhaps?

**Comment by bojanz** [comment]
Btw, what's your refresh_frequency set to? The default is 30s, which I figure would be enough to hide this problem on most installs. I wonder if that check in needsRefresh() is somehow bypassed?

File: modules/order/src/OrderStorage.php (hunk 3 above)
Comment (agoradesign) [reply modules/order/src/OrderStorage.php:54]: ok, I'll change that later this day

**Comment by agoradesign** [comment]
That's a good point... Apparently I haven't taken care of this settings after updating to beta... So it's currently set to nothing :D

I'll try that with the default setting. Of course it should work with a zero value as well, but it makes the issue far less critical then

PS: I've set a refresh frequency, and the issue doesn't happen any more. Even a second should be enough during a single request

**Comment by bojanz** [comment]
Committed in 0dfb84b0e335054470d5a4dda7bc90303e33f89a.

**Pull request closed without merging**